File: LeetCode/17.letter-combinations-of-a-phone-number.cpp
/*
 * @lc app=leetcode id=17 lang=cpp
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
class Solution {
public:
   
    vector<string> letterCombinations(string digits) {


        vector<string> str = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv","wxyz"};


        vector<string> ans;

        int n=digits.size();

        if(n==0)
            return ans;


        rec(str,ans,digits,0,digits.size(),"");

        return ans;



       

       


    }

    void rec(vector<string> &str,vector<string> &ans,string digits,int i,int n,string temp)
    {
        if(i==n)
        {
            ans.push_back(temp);
            return;
        }

        int index=digits[i]-'0';

        for(int j=0;j<str[index].size();j++)
        {
            temp.push_back(str[index][j]);
            rec(str,ans,digits,i+1,n,temp);
            temp.pop_back();
        }
    }



    



};
// @lc code=end

File: LeetCode/17.letter-combinations-of-a-phone-number.py
#
# @lc app=leetcode id=17 lang=python3
#
# [17] Letter Combinations of a Phone Number
#

# @lc code=start
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:

        if not digits:
            return []

        phone = {
            '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
            '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
        }

        def backtrack(index, path):
            if len(path) == len(digits):
                result.append(''.join(path))
                return

            for i in range(index, len(digits)):
                for letter in phone[digits[i]]:
                    path.append(letter)
                    backtrack(i + 1, path)
                    path.pop()

        result = []
        backtrack(0, [])
        return result

        



        
        
# @lc code=end

